ntoskrnl/ex/locale.c

```diff
@@ -24,8 +24,252 @@ LANGID PsInstallUILanguageId = LANGIDFROMLCID(0x00000409);
+    LocaleName.Buffer = LocalBuffer;
+
+    /* Convert the locale ID into a suitable string */
+    Status = RtlIntegerToUnicodeString(LocaleId, 16, &LocaleName);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to convert LCID to a string (Status 0x%lx)\n", Status);
+        return Status;
+    }
+
+    /* Format the LCID buffer into a hexadecimal format */
+    HexId = ValueIdBuffer;
+    IdLength = LocaleName.Length / sizeof(WCHAR);
+    for (i = IdLength; i < LCID_HEX_FORMAT_DIGIT; i++)
+    {
+        *HexId = L'0';
+        HexId++;
+    }
+
+    /* Null-terminate the buffer and compute the length */
+    *HexId = UNICODE_NULL;
+    BufferLength = sizeof(ValueIdBuffer) / sizeof(WCHAR);
+
+    /* Finally append the actual LCID string into the formatted hexadecimal string */
+    RtlInitUnicodeString(&LocaleIdString, ValueIdBuffer);
+    LocaleIdString.MaximumLength = BufferLength;
+    RtlAppendUnicodeToString(&LocaleIdString, LocaleName.Buffer);
+
+    /* Open the NLS locale key */
+    Status = ZwOpenKey(&NlsLocaleKey,
+                       KEY_QUERY_VALUE,
+                       &NlsLocalKeyAttrs);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to open %wZ (Status 0x%lx)\n", NlsLocaleKeyPath, Status);
+        return Status;
+    }
+
+    /* Open the NLS alternate sort locales key */
+    Status = ZwOpenKey(&AltSortKey,
+                       KEY_QUERY_VALUE,
+                       &AltSortKeyAttrs);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to open %wZ (Status 0x%lx)\n", AltSortKeyPath, Status);
+        goto Quit;
+    }
+
+    /* Open the NLS language groups key */
+    Status = ZwOpenKey(&LangGroupKey,
+                       KEY_QUERY_VALUE,
+                       &LangGroupKeyAttrs);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to open %wZ (Status 0x%lx)\n", LangGroupPath, Status);
+        goto Quit;
+    }
+
+    /* Check if the captured locale ID exists in the list of other locales */
+    BufferKey = (PKEY_VALUE_PARTIAL_INFORMATION)ValueBuffer;
+    Status = ZwQueryValueKey(NlsLocaleKey,
+                             &LocaleIdString,
+                             KeyValuePartialInformation,
+                             BufferKey,
+                             sizeof(ValueBuffer),
+                             &ReturnedLength);
+    if (!NT_SUCCESS(Status))
+    {
+        /* We failed, retry by looking at the alternate sorts locales */
+        Status = ZwQueryValueKey(AltSortKey,
+                                 &LocaleIdString,
+                                 KeyValuePartialInformation,
+                                 BufferKey,
+                                 sizeof(ValueBuffer),
+                                 &ReturnedLength);
+        if (!NT_SUCCESS(Status))
+        {
+            DPRINT1("Failed to query value from Alternate Sorts key (Status 0x%lx)\n", Status);
+            goto Quit;
+        }
+    }
```

Comment:
Could you check, there, that the data type and possibly, length, is what's expected? (should it be REG_SZ ?)